File: lib/Option/Options.cpp
//===--- Options.cpp - Option info & table --------------------------------===//
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
//
//===----------------------------------------------------------------------===//

#include "swift/Option/Options.h"

#include "llvm/ADT/STLExtras.h"
#include "llvm/Option/OptTable.h"
#include "llvm/Option/Option.h"

#define OPTTABLE_STR_TABLE_CODE
#include "swift/Option/Options.inc"
#undef OPTTABLE_STR_TABLE_CODE

#define OPTTABLE_PREFIXES_TABLE_CODE
#include "swift/Option/Options.inc"
#undef OPTTABLE_PREFIXES_TABLE_CODE

using namespace swift::options;
using namespace llvm::opt;

static const llvm::opt::GenericOptTable::Info InfoTable[] = {
#define OPTION(...) LLVM_CONSTRUCT_OPT_INFO(__VA_ARGS__),
#include "swift/Option/Options.inc"
#undef OPTION
};

namespace {

class SwiftOptTable : public llvm::opt::GenericOptTable {
public:
  SwiftOptTable()
      : GenericOptTable(OptionStrTable, OptionPrefixesTable, InfoTable) {}
};

} // end anonymous namespace

std::unique_ptr<OptTable> swift::createSwiftOptTable() {
  return std::unique_ptr<GenericOptTable>(new SwiftOptTable());
}

File: SwiftCompilerSources/Sources/Optimizer/PassManager/Options.swift
//===--- Options.swift ----------------------------------------------------===//
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
//
//===----------------------------------------------------------------------===//

import SIL
import OptimizerBridging

struct Options {
  let _bridged: BridgedPassContext

  var enableStackProtection: Bool {
    _bridged.enableStackProtection()
  }

  var useAggressiveReg2MemForCodeSize : Bool {
    _bridged.useAggressiveReg2MemForCodeSize()
  }

  var enableMoveInoutStackProtection: Bool {
    _bridged.enableMoveInoutStackProtection()
  }

  func enableSimplification(for inst: Instruction) -> Bool {
    _bridged.enableSimplificationFor(inst.bridged)
  }

  func enableAddressDependencies() -> Bool {
    _bridged.enableAddressDependencies()
  }

  var noAllocations: Bool {
    _bridged.noAllocations()
  }

  var enableEmbeddedSwift: Bool {
    hasFeature(.Embedded)
  }

  var enableMergeableTraps: Bool {
    _bridged.enableMergeableTraps()
  }

  func hasFeature(_ feature: BridgedFeature) -> Bool {
    _bridged.hasFeature(feature)
  }

  // The values for the assert_configuration call are:
  // 0: Debug
  // 1: Release
  // 2: Fast / Unchecked
  enum AssertConfiguration {
    case debug
    case release
    case unchecked
    case unknown

    var integerValue: Int {
      switch self {
      case .debug:      return 0
      case .release:    return 1
      case .unchecked:  return 2
      case .unknown:    fatalError()
      }
    }
  }

  var assertConfiguration: AssertConfiguration {
    switch _bridged.getAssertConfiguration() {
      case .Debug:               return .debug
      case .Release:             return .release
      case .Unchecked:           return .unchecked
      default:                   return .unknown
    }
  }
}
